File: src/main/js/g0001_0100/s0025_reverse_nodes_in_k_group/solution.js
// #Hard #Top_100_Liked_Questions #Linked_List #Recursion #Data_Structure_II_Day_13_Linked_List
// #Udemy_Linked_List #Top_Interview_150_Linked_List #Big_O_Time_O(n)_Space_O(k)
// #2024_12_04_Time_0_ms_(100.00%)_Space_53.3_MB_(54.66%)

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
const reverseKGroup = function (head, k) {
    if (head === null || head.next === null || k === 1) {
        return head
    }

    let len = head
    let count = 0

    while (count < k) {
        if (len === null) {
            return head
        }
        len = len.next
        count++
    }

    let current = head
    let next = null
    let prev = null
    let i = 0

    while (i < k) {
        next = current.next
        current.next = prev
        prev = current
        current = next
        i++
    }

    head.next = reverseKGroup(next, k)

    return prev
};

export { reverseKGroup }

File: src/main/dart/g0001_0100/s0025_reverse_nodes_in_k_group/Solution.dart
// #Hard #Top_100_Liked_Questions #Linked_List #Recursion #Data_Structure_II_Day_13_Linked_List
// #Udemy_Linked_List #Top_Interview_150_Linked_List #Big_O_Time_O(n)_Space_O(k)
// #2024_10_01_Time_357_ms_(73.33%)_Space_153.5_MB_(6.67%)

/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
  ListNode? reverseKGroup(ListNode? head, int k) {
    if (head == null || head.next == null || k == 1) {
      return head;
    }

    int j = 0;
    ListNode? len = head;
    // Check if the linked list has fewer than `k` nodes, return the list as is.
    while (j < k) {
      if (len == null) {
        return head;
      }
      len = len.next;
      j++;
    }

    // Reverse `k` nodes.
    ListNode? current = head;
    ListNode? next;
    ListNode? prev = null;
    int i = 0;

    // Traverse and reverse the nodes in groups of `k`.
    while (i < k) {
      next = current!.next;
      current.next = prev;
      prev = current;
      current = next;
      i++;
    }

    // Recursively reverse the next `k` group of nodes.
    head!.next = reverseKGroup(current, k);

    // Return the new head of the reversed list.
    return prev;
  }
}
